File: Case Maroon_Unity/Assets/Scripts/CaseMaroon/GameSystems/Player.cs

using UnityEngine;
using System.Collections.Generic;
using static CaseMaroon.Miscellaneous.GlobalData;
using static CaseMaroon.Backend.BackendResponses;
using System.Linq;
using Unity.VisualScripting;
using static CaseMaroon.Backend.BackendModels;

namespace CaseMaroon.GameSystem
{
    public class Player
    {
        public int Id { get; set; }

        public HashSet<ReconPosition> ReconPositions { get; private set; }

        public HashSet<Vector2Int> OwnedPositions { get; private set; }

        public Player(PlayerModel pr)
        {
            Id = pr.id;

            ReconPositions = new HashSet<ReconPosition>();
            OwnedPositions = new HashSet<Vector2Int>();

            if(pr.reconPositions.Count > 0)
            {
                ReconPositions.AddRange(pr.reconPositions);
            }
        }

        public void UpdateReconPosition(List<ReconPosition> add, List<ReconPosition> remove)
        {
            AddReconPositions(add);
            RemoveReconPositions(remove);
        }

        private void AddReconPosition(ReconPosition reconPos)
        {
            if (ReconPositions.TryGetValue(reconPos, out var existing))
            {
                ReconPositions.Remove(existing);
                existing.AddRecon(reconPos);
                ReconPositions.Add(existing);
            }
            else
            {
                ReconPositions.Add(reconPos);
            }
        }
        private void RemoveReconPosition(ReconPosition reconPos)
        {
            if (ReconPositions.TryGetValue(reconPos, out var existing))
            {
                ReconPositions.Remove(existing);
                existing.RemoveRecon(reconPos);

                // Optionally remove if recon drops to 0
                if (existing.ReconLevel > 0)
                    ReconPositions.Add(existing);
            }
        }
        private void AddReconPositions(List<ReconPosition> reconPositions)
        {
            foreach (var reconPos in reconPositions)
            {
                AddReconPosition(reconPos);
            }
        }

        private void RemoveReconPositions(List<ReconPosition> reconPositions)
        {
            foreach (var reconPos in reconPositions)
            {
                RemoveReconPosition(reconPos);
            }
        }
        private void RequestDataFromSever()
        {
            // when the game starts, request data such as positions, units, building...
            // to be implemented
        }

        public void Update(PlayerModel pr)
        {
            // Update the player's data with the server data
            Id = pr.id;
            // Update recon positions
            ReconPositions.Clear();
            ReconPositions.UnionWith(pr.reconPositions);
        }

        public void SetOwnedPositions(List<OwnedPosition> positions)
        {
            List<Vector2Int> pos = positions.Where(pos => pos.playerId == Id).Select(pos => pos.gridPosition).ToList();

            OwnedPositions.Clear();
            OwnedPositions.UnionWith(pos);
        }

        public bool CanSeeUnit(Vector2Int gridPos)
        {
            // Check if the player can see a unit at the given grid position
            return ReconPositions.Any(rp => rp.gridPosition == gridPos && rp.ReconLevel > 2);
        }

    }
}

File: Case Maroon_Backend/src/models/GameSystems/Player.js
import {
  ReconPosition,
  ReconPositionSchema,
  Vector2IntSchema,
} from "../Miscellaneous/index.js";

export class Player {
  static nextId = 1;
  static reconScope = 3;
  static reconLevel = 2;

  constructor(pid, username) {
    this.id = pid;
    this.username = username;
    
    this.reconPositions = []; // Array of ReconPosition

    this.initState = true;
  }


  // Helper to check if a position exists in an array
  _findReconIndex(gridPosition) {
    return this.reconPositions.findIndex(
      (rp) =>
        rp.gridPosition.x === gridPosition.x &&
        rp.gridPosition.y === gridPosition.y
    );
  }

  _findOwnedIndex(gridPosition) {
    return this.ownedPositions.findIndex(
      (pos) => pos.x === gridPosition.x && pos.y === gridPosition.y
    );
  }

  addReconPosition(reconPos) {
    const idx = this._findReconIndex(reconPos.gridPosition);

    if (idx !== -1) {
      this.reconPositions[idx].addRecon(reconPos);
    } else {
      this.reconPositions.push(reconPos);
    }
  }

  removeReconPosition(reconPos) {
    const idx = this._findReconIndex(reconPos.gridPosition);

    if (idx !== -1) {
      this.reconPositions[idx].removeRecon(reconPos);

      if (this.reconPositions[idx].reconLevel <= 0) {
        this.reconPositions.splice(idx, 1);
      }
    }
  }

  addReconPositions(reconPositions) {
    for (const rp of reconPositions) {
      this.addReconPosition(rp);
    }
  }

  removeReconPositions(reconPositions) {
    for (const rp of reconPositions) {
      this.removeReconPosition(rp);
    }
  }
}

export const PlayerState = {
    0: "Loading",
    1: "Idle",
    2: "WaitingForHQPlacement",
    3: "WaitingForUnitPlacement",
    4: "WaitingForNext",
    
    Loading: 0,
    Idle: 1,
    WaitingForHQPlacement: 2,
    WaitingForUnitPlacement: 3,
    WaitingForNext: 4
};